src/decoder/mod.rs

```diff
@@ -54,7 +64,50 @@ where
     Mp3(mp3::Mp3Decoder<R>),
     #[cfg(feature = "symphonia")]
     Symphonia(symphonia::SymphoniaDecoder),
-    None(::std::marker::PhantomData<R>),
+    None(std::marker::PhantomData<R>),
+}
+
+pub struct DecoderInfo {
+    channels: u16,
+    sample_rate: u32,
+    samples_elapsed: AtomicUsize,
+}
+
+impl DecoderInfo {
+    fn from_source<S>(source: &S) -> Self
+    where
+        S: Source,
+        S::Item: Sample,
+    {
+        Self {
+            channels: source.channels(),
+            sample_rate: source.sample_rate(),
+            samples_elapsed: AtomicUsize::default(),
+        }
+    }
+
+    #[inline]
+    pub fn channels(&self) -> u16 {
+        self.channels
+    }
+
+    #[inline]
+    pub fn sample_rate(&self) -> u32 {
+        self.sample_rate
+    }
+
+    #[inline]
+    pub fn elapsed_samples(&self) -> usize {
+        self.samples_elapsed.load(Ordering::Relaxed)
+    }
+
+    #[inline]
+    pub fn elapsed_duration(&self) -> Option<Duration> {
+        Duration::try_from_secs_f64(
```

Comment:
forgive me but why do we return an `Option` here? is there a combination of elapsed_samples, sample rate and channels that makes [try_from_secs_f64](https://doc.rust-lang.org/std/time/struct.Duration.html#method.try_from_secs_f64) fail? I would love a comment explaining that.

Or is it there because some combination of sources could not support `elapsed_duration`?